dss_metrics/minio_rest_collector.py

```diff
@@ -0,0 +1,132 @@
+NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import json
+import metrics
+import re
+import requests
+import socket
+import subprocess
+import time
+
+
+class MinioRESTCollector(object):
+    def __init__(self, configs, whitelist_patterns, filter=False):
+        self.configs = configs
+        self.whitelist_patterns = whitelist_patterns
+        self.filter = filter
+        self.minio_metrics = {'minio_disk_storage_used_bytes'}
+        self.metrics_url_prefix = "http://"
+        self.metrics_url_suffix = "/minio/prometheus/metrics"
+        self.mc = configs['mc_binary_path']
+        self.conf_json_bucket_suffix = configs['conf_json_bucket_suffix']
+        self.cluster_id = self.configs['cluster_id']
+        self.TYPE = 'minio'
+
+    def poll_statistics(self, metrics_data_buffer):
+        cluster_endpoint_map = self.get_miniocluster_endpoint_map()
+        for id, endpts in cluster_endpoint_map.items():
+            miniocluster_id = id
+            minio_endpoint = endpts.pop()
+            minio_metrics = self.get_minio_metrics_from_endpoint(
+                minio_endpoint)
+
+            tags = {}
+            tags['cluster_id'] = self.cluster_id
+            tags['target_id'] = socket.gethostname()
+            tags['minio_id'] = miniocluster_id
+            tags['type'] = self.TYPE
+
+            for metric in minio_metrics:
+                if self.filter and not self.check_whitelist_key(metric[0]):
+                    continue
+                metrics_data_buffer.append(
+                    metrics.MetricInfo(
+                        metric[0], metric[0], metric[1], tags, time.time())
+                )
+
+    def check_whitelist_key(self, key):
+        for regex in self.whitelist_patterns:
+            if re.match(regex, key):
+                return True
+        return False
+
+    def get_miniocluster_endpoint_map(self):
+        get_hosts_cmd = self.mc + " " + "config host list"
+        proc = subprocess.Popen(get_hosts_cmd.split(' '),
+                                stdout=subprocess.PIPE)
+        local_minio_host = None
+        miniocluster_endpoint_map = {}  # {<miniocluster> : {<endpoints>} }
+
+        for line in proc.stdout.readlines():
+            # find local minio host or endpoint
+            decoded_line = line.decode('utf-8').strip("\n")
+            if decoded_line.startswith('local_'):
```

Comment:
using "local_" is not a good practice. Ask Matt to add Samsung or some other prefix to identify our own endpoints.

Reply:
Formulating plan with Matt to deploy Metrics Agent with ansible, which will then enable metrics agent to have access to MINIO metrics. Once that is done, then I can use S3 API instead of mc to get conf.json data

Reply:
for now Matt will not change endpoint names, hence I need to shift over to S3 API to retrieve conf.json file. I will open a separate PR for this once we have added support to deploy Metrics Agent along with DSS